### ISSUE TITLE
CVE-2022-1471 - SnakeYAML - Arbitrary Code Execution

### ISSUE THREAD

**Issue by gaetanmaisse** [gravitee-io/issues#9419]
TLDR;

APIM is not affected by CVE-2022-1471 .

## Direct use of SnakeYAML (in APIM codebase)

The SnakeYAML library is used in the APIM codebase to parse YAML files. It is directly used in a single place: [https://github.com/gravitee-io/gravitee-api-management/blob/3.20.25/gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/NotificationTemplateServiceImpl.java#L298-L299](https://github.com/gravitee-io/gravitee-api-management/blob/3.20.25/gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/NotificationTemplateServiceImpl.java#L298-L299)

This is an extract of the code:

<pre>...
 final String RELATIVE_TPL_PATH = "notifications/portal/";
 String templateName = hook.getTemplate() + "." + io.gravitee.rest.api.model.notification.NotificationTemplateType.PORTAL.name();
 File portalTemplateFile = new File(templatesPath + "/" + RELATIVE_TPL_PATH + hook.getTemplate() + ".yml");

...
 Yaml yaml = new Yaml();
 Map
 load = yaml.load(new FileInputStream(portalTemplateFile));
 </pre>

This Java method is part of the NotificationTemplateServiceImpl class, and it's designed to load a YAML formatted notification template file from a file system, parse it, and then return a new NotificationTemplateEntity object that has been populated using values from the YAML file.

It will load the following YAML files `

 {gravitee.home} 

/templates/notifications/portal/**.yml`, these files are used to define the content of the email notifications that are sent to users when they register to the portal, when they reset their password, etc. They are defined on the instance running the Management API of Gravitee APIM and so are not accessible nor modifiable externally.

**Conclusion**

**The attacker needs to have access to the file system of Management API's instance to exploit it.** If the attacker has access to it, this CVE is the least of your problems.

## Indirect use of SnakeYAML (via dependencies)

SnakeYAML is also used indirectly via dependencies, and it's the case for the following dependencies:

*   jackson-databind
*   jackson-dataformat-yaml
*   swagger-parser
*   swagger-core
*   gravitee-node

However, none of these are vulnerable:

*   jackson-databind: [https://github.com/FasterXML/jackson-databind/issues/3697](https://github.com/FasterXML/jackson-databind/issues/3697)
*   jackson-dataformat-yaml: [https://github.com/FasterXML/jackson-dataformats-text/issues/361#issuecomment-1350117166](https://github.com/FasterXML/jackson-dataformats-text/issues/361#issuecomment-1350117166)

> Jackson YAML module does not use SnakeYAML's databinding level for anything – so neither Constructor nor its subtypes are used for anything. Only token-/event-stream decoding (and reference resolution) is used.

*   swagger-parser: NA

> In swagger-parser we are using SafeConstructor which is the way to avoid the vulnerabilty, therefore Swagger Parser is not affected by the vulnerabilty

*   swagger-core: [https://github.com/swagger-api/swagger-core/issues/4323#issuecomment-1351815336](https://github.com/swagger-api/swagger-core/issues/4323#issuecomment-1351815336)

> In swagger-core Jackson library (which has transitive snakeyaml library) has an open ticket which we are monitoring, it also has another ticket related to possibly the same kind of vulnerability, ticket closed as usage is not affected by vulnerability.

*   gravitee-node: use SnakeYAML indirectly via `spring-beans` (more precisely the `YamlProcessor` class) [https://github.com/spring-projects/spring-framework/issues/29649#issuecomment-1339765926](https://github.com/spring-projects/spring-framework/issues/29649#issuecomment-1339765926)

> Spring Framework's YamlProcessor already restricts the types allowed to be loaded via its org.springframework.beans.factory.config.YamlProcessor#setSupportedTypes method. By default, only Java standard classes (as defined in SafeConstructor are allowed). With this limitation in place, I don't think the security issue you're pointing to can be leveraged.

**Conclusion**

**None of the dependencies transitively using SnakeYAML are vulnerable.**
